server/datastore/mysql/mdm_test.go

```diff
@@ -1276,3 +1278,167 @@ func testBulkSetPendingMDMHostProfiles(t *testing.T, ds *Datastore) {
+			ExecAdhocSQL(t, ds, func(tx sqlx.ExtContext) error {
+				var got []fleet.ConfigurationProfileLabel
+				err := sqlx.SelectContext(ctx, tx, &got, query, args...)
+				require.NoError(t, err)
+				require.Len(t, got, len(want))
+				return nil
+			})
+		}
+
+		t.Run("empty input "+platform, func(t *testing.T) {
+			want := []fleet.ConfigurationProfileLabel{}
+			err := ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+				return batchSetProfileLabelAssociationsDB(ctx, tx, want, platform)
+			})
+			require.NoError(t, err)
+			expectLabels(t, want)
+		})
+
+		t.Run("valid input "+platform, func(t *testing.T) {
+			profileLabels := []fleet.ConfigurationProfileLabel{
+				{ProfileUUID: uuid, LabelName: label.Name, LabelID: label.ID},
+			}
+			err := ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+				return batchSetProfileLabelAssociationsDB(ctx, tx, profileLabels, platform)
+			})
+			require.NoError(t, err)
+			expectLabels(t, profileLabels)
+		})
+
+		t.Run("invalid profile UUID "+platform, func(t *testing.T) {
+			invalidProfileLabels := []fleet.ConfigurationProfileLabel{
+				{ProfileUUID: "invalid-uuid", LabelName: label.Name, LabelID: label.ID},
+			}
+
+			err := ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+				return batchSetProfileLabelAssociationsDB(ctx, tx, invalidProfileLabels, platform)
+			})
+			require.Error(t, err)
+		})
+
+		t.Run("invalid label data "+platform, func(t *testing.T) {
+			// invalid id
+			invalidProfileLabels := []fleet.ConfigurationProfileLabel{
+				{ProfileUUID: uuid, LabelName: label.Name, LabelID: 12345},
+			}
+			err := ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+				return batchSetProfileLabelAssociationsDB(ctx, tx, invalidProfileLabels, platform)
+			})
+			require.Error(t, err)
+
+			// both invalid
+			invalidProfileLabels = []fleet.ConfigurationProfileLabel{
+				{ProfileUUID: uuid, LabelName: "xyz", LabelID: 1235},
+			}
+			err = ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+				return batchSetProfileLabelAssociationsDB(ctx, tx, invalidProfileLabels, platform)
+			})
+			require.Error(t, err)
+		})
+
+		t.Run("labels are removed "+platform, func(t *testing.T) {
+			// create a new label
+			newLabel := &fleet.Label{
+				Name:        "new label" + platform,
+				Description: "a label",
+				Query:       "select 1 from orbit_info;",
+			}
+			newLabel, err := ds.NewLabel(ctx, newLabel)
+			require.NoError(t, err)
+
+			// apply a batch set with the new label
+			profileLabels := []fleet.ConfigurationProfileLabel{
+				{ProfileUUID: uuid, LabelName: label.Name, LabelID: label.ID},
+				{ProfileUUID: uuid, LabelName: newLabel.Name, LabelID: newLabel.ID},
+			}
+			err = ds.withTx(ctx, func(tx sqlx.ExtContext) error {
+				return batchSetProfileLabelAssociationsDB(ctx, tx, profileLabels, platform)
+			})
+			require.NoError(t, err)
+			// both are stored in the DB
+			expectLabels(t, profileLabels)
+
+			// batch apply again without the label
```

Comment:
I think that's a bit clearer?

```suggestion
			// batch apply again without the newLabel
```